src/models/tree/build_tree.cu

```diff
@@ -107,32 +106,41 @@ class GradientQuantiser {
   {
     thrust::default_random_engine eng(seed);
     thrust::uniform_real_distribution<double> dist(0.0, 1.0);
-    auto scaled_grad                        = value.grad * scale.grad;
-    auto scaled_hess                        = value.hess * scale.hess;
-    double grad_remainder                   = scaled_grad - floor(scaled_grad);
-    double hess_remainder                   = scaled_hess - floor(scaled_hess);
+    auto scaled_grad      = value.grad * scale.grad;
+    auto scaled_hess      = value.hess * scale.hess;
+    double grad_remainder = scaled_grad - floor(scaled_grad);
+    double hess_remainder = scaled_hess - floor(scaled_hess);
+    // We won't check for overflow here as this is performance critical
+    // If our calculation of the scale factor is correct we should never overflow
+    // NOLINTBEGIN(cppcoreguidelines-narrowing-conversions)
     IntegerGPair::value_type grad_quantised = floor(scaled_grad) + (dist(eng) < grad_remainder);
     IntegerGPair::value_type hess_quantised = floor(scaled_hess) + (dist(eng) < hess_remainder);
```

Comment:
Narrow activately checks for overflow and can __trap() the cuda kernel, but this has a performance penalty and this section is important for performance. I am instead relying on the math we did previously so that it doesn't overflow.